Add meta schema: a JSON Schema which matches recap schemas

mjperrone opened the pull request:
## What

This adds a [JSON Schema](https://json-schema.org/) describing valid recap schemas. 

## Why

There is a robust ecosystem around JSON Schema that provides tools like validation, example generation, code generation, etc out of the schema definitions. Defining a JSON Schema definition for Recap will allow Recap and its users to more easily hook into that ecosystem. 

Closes https://github.com/recap-build/recap/issues/342

## How

I manually created the JSON Schema file by reading closely the [Recap Spec](https://recap.build/specs/type/0.2.0/).

## Known limitations

As mentioned in the issue, I don't know of a way to require that the `alias` strings are referencing defined types elsewhere in the document, so some invalid documents will pass. No documents that are valid should fail.

## Validation

In order to help iterate on the JSON Schema definition (which I expect isn't perfect), I test up some unit tests. I define a few valid recap schema documents and use [jsonschema](https://pypi.org/project/jsonschema/) to test the documents against the schema definition. 

## Potential future work

* Define a meta schema in recap (maybe using recap converter from the json schema)
* flesh out more positive or negative examples
* use the JSON Schema definition for something like editor integrated validation

mjperrone commented on tests/spec/invalid_field_type.json:
```diff
@@ -0,0 +1,8 @@
+    {
+    "name": "string_field",
+    "doc": "doc",
+    "type": "invalidtype",
```
I included this in the initial draft PR to show what the failing test UX is like:

```
E           jsonschema.exceptions.ValidationError: {'name': 'string_field', 'doc': 'doc', 'type': 'invalidtype', 'bytes': 255} is not valid under any of the given schemas
E           
E           Failed validating 'oneOf' in schema['items']:
E               {'oneOf': [{'allOf': [{'$ref': '#/$defs/RecapNull'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapBool'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapInt'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapFloat'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapString'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapBytes'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapList'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapMap'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapStruct'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapEnum'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]},
E                          {'allOf': [{'$ref': '#/$defs/RecapUnion'},
E                                     {'$ref': '#/$defs/RecapCommonFields'}]}]}
E           
E           On instance[0]:
E               {'bytes': 255,
E                'doc': 'doc',
E                'name': 'string_field',
E                'type': 'invalidtype'}

.venv/lib/python3.10/site-packages/jsonschema/validators.py:1302: ValidationError
==================================================================== short test summary info =====================================================================
FAILED tests/unit/meta_schema/test_json_schema.py::test_valid_schemas[tests/spec/invalid_field_type.json] - jsonschema.exceptions.ValidationError: {'name': 'string_field', 'doc': 'doc', 'type': 'invalidtype', 'bytes': 255} is not valid under any of the given schemas
```

mjperrone replied:
Updated the harness to find `.unimplemented.invalid.json` files and expect that they will throw as invalid schemas, but mark that test as expected to fail while unimplemented.

mjperrone commented on tests/0.2.0.json:
```diff
@@ -0,0 +1,377 @@
+{
```
After this JSON Schema is reviewed and approved in this PR, I will remove this from this PR and add it to `recap-website/specs/type/0.2.0.json`, then switch from a file reference to a network call.

I'm including it in this PR draft to make it easier to iterate on the spec and tests in parallel.

mjperrone commented on tests/0.2.0.json:
```diff
@@ -0,0 +1,377 @@
+                "$ref": "#/$defs/RecapNull"
+              },
+              {
+                "$ref": "#/$defs/RecapCommonFields"
```
I chose to put the common fields `alias` and `doc` in its own def instead of repeating them in each of the base types (like `RecapNull` here).

I also chose to put the `allOf` in this `RecapType` `oneOf` union instead of inside each of the base types. I did this both to mirror the DRY way the existing spec introduces these fields, and to emphasize the unique fields when looking at the base types.

mjperrone replied:
I ended up repeating myself so that the error messages are easier to read when validation fails.

mjperrone commented on tests/0.2.0.json:
```diff
@@ -0,0 +1,377 @@
+          },
+          "bits": {
+            "type": "integer",
+            "minimum": 1,
+            "maximum": 64
```
I'm not sure about the correct ranges for some of these integer values

criccomini replied:
Integer `bits` field is a 32-bit signed int, so min is 1 and max is 2_147_483_647.

Same for `bits` in float.

`bytes` for string and bytes is 64-bit signed. The `attributes` section in the spec has some info there.

mjperrone replied:
updated max for these fields

mjperrone commented on tests/0.2.0.json:
```diff
@@ -0,0 +1,377 @@
+          }
+        }
+      }
+    }
```
I don't think I have represented Logical types in this

mjperrone replied:
I've done 1 and made a note to make an issue for 2 after this PR is finished

criccomini commented:
One other thought:

I could see an argument for being more explicit about the recap types (removing allOf/oneOf), even if it means repeating yourself. The argument for this would be that it makes the JSON schema way more readable to humans.

I'm indifferent on that argument at this point, but I could see the spec being a bit confusing to humans that are trying to read (vs. validators like `python-jsonschema`). I was a bit disoriented when I looked at it; I didn't actually understand until you mentioned the allOf/oneOf in the comment.

criccomini commented:
Re: the `black` failure, try running `pdm style` to get things cleaned up.

mjperrone commented:
> One thing I didn't catch that I think is a common gotcha: disallowing `alias` to be set when `type` reference an alias.

Did this by not repeating the `alias` definition for `RecapAliasReference`

> I could see an argument for being more explicit about the recap types (removing allOf/oneOf), even if it means repeating yourself. The argument for this would be that it makes the JSON schema way more readable to humans.

I ended up doing as I was iterating on this because the validation errors were harder to understand.

criccomini reviewed (changes requested):
🔥 This looks great.

Only request:

1. Can you move unit/meta_schema/test_json_schema.py to tests/spec
2. And then move spec/*.json to a subdir like `schemas` or something?
3. And then update the [.github/workflows/ci.yaml](https://github.com/recap-build/recap/blob/main/.github/workflows/ci.yaml) to have an extra job called `spec-tests` that runs `pdm run pytest tests/spec -vv`

I want to keep the spec enforcement separate from the standard unit tests.

mjperrone commented:
> 1. Can you move unit/meta_schema/test_json_schema.py to tests/spec
> 2. And then move spec/*.json to a subdir like `schemas` or something?
> 3. And then update the [.github/workflows/ci.yaml](https://github.com/recap-build/recap/blob/main/.github/workflows/ci.yaml) to have an extra job called `spec-tests` that runs `pdm run pytest tests/spec -vv`

1. done
2. moved them to `spec/{in,}valid`, so removed the `.invalid.` file sub-extension
3. done. I just followed the pattern of the existing jobs, but it's doing checkout+python+pdm installs for each step. That could be cached so it only has to happen once per build, or really once per new dependencies

I also realized that these tests should test that recap is validating the schemas the same way as the JSON Schema, so I added the recap validation to each.

This made me realize that there's no reason we shouldn't take every positive and negative validation example in the recap test suites and evaluate that against the jsonschema validation as well. Given the current structure of the tests, how feasible does that sound?

criccomini commented:
💥 Merged! Thanks!!

Follow-on things I found:

1. Migrating .json spec to recap-website
2. Update to pull spec from the recap-website
3. GH issue for logical type spec tests
4. (Maybe?) GH issue to do validation when translating to/from recap CST for proto/avro/JSON schema.

mjperrone commented:
> 💥 Merged! Thanks!!
> 
> Follow-on things I found:
> 
> 1. Migrating .json spec to recap-website
> 2. Update to pull spec from the recap-website
> 3. GH issue for logical type spec tests
> 4. (Maybe?) GH issue to do validation when translating to/from recap CST for proto/avro/JSON schema.

3. https://github.com/recap-build/recap/issues/348

Another thought: this test suite could be extended to read specs defined in YAML or TOML. There ¿might? be some situations where those languages handle values differently like `null`

criccomini commented:
Yup, totally. I think it's worth defining in YAML, TOML, and JSON, and validating the CST using the python-jsonschema lib.

One gotcha with TOML, in particular, is that it doesn't have a null AFAIK. Makes null defaults tricky. At least we added `optional` for the common case, which will work with TOML.

criccomini commented:
Opened these follow-on issues:

https://github.com/recap-build/recap/issues/349
https://github.com/recap-build/recap/issues/350
https://github.com/recap-build/recap/issues/351
